waspc/packages/wasp-config/src/mapUserSpecToAppSpecDecls.ts

```diff
@@ -64,60 +71,39 @@ export function mapUserSpecToDecls(
       emailSender,
       websocket
     ),
-  })
-
-  for (const [actionName, actionConfig] of actions.entries()) {
-    decls.push({
-      declType: 'Action',
-      declName: actionName,
-      declValue: mapOperationConfig(actionConfig, parseEntityRef),
-    })
-  }
-
-  for (const [queryName, queryConfig] of queries.entries()) {
-    decls.push({
-      declType: 'Query',
-      declName: queryName,
-      declValue: mapOperationConfig(queryConfig, parseEntityRef),
-    })
   }
 
-  for (const [apiName, apiConfig] of apis.entries()) {
-    decls.push({
-      declType: 'Api',
-      declName: apiName,
-      declValue: mapApiConfig(apiConfig, parseEntityRef),
-    })
-  }
-
-  for (const [jobName, jobConfig] of jobs.entries()) {
-    decls.push({
-      declType: 'Job',
-      declName: jobName,
-      declValue: mapJob(jobConfig, parseEntityRef),
-    })
-  }
-
-  for (const [
-    apiNamespaceName,
-    apiNamespaceConfig,
-  ] of apiNamespaces.entries()) {
-    decls.push({
-      declType: 'ApiNamespace',
-      declName: apiNamespaceName,
-      declValue: mapApiNamespace(apiNamespaceConfig),
-    })
-  }
+  return makeDeclsArray({
+    App: [appDecl],
+    Page: pageDecls,
+    Route: routeDecls,
+    Action: actionDecls,
+    Query: queryDecls,
+    Api: apiDecls,
+    Job: jobDecls,
+    ApiNamespace: apiNamespaceDecls,
+    Crud: crudDecls,
+  })
+}
 
-  for (const [crudName, crudConfig] of spec.cruds.entries()) {
-    decls.push({
-      declType: 'Crud',
-      declName: crudName,
-      declValue: mapCrud(crudConfig, parseEntityRef),
-    })
-  }
+function makeDeclsArray(decls: {
+  [Type in AppSpec.Decl['declType']]: AppSpec.GetDeclForType<Type>[]
+}): AppSpec.Decl[] {
+  return Object.values(decls).flatMap((decl) => [...decl])
+}
 
-  return decls
+function mapToDecls<T, const DeclType extends AppSpec.Decl['declType']>(
+  configs: Map<string, T>,
+  type: DeclType,
+  configToDeclValue: (
+    config: T
+  ) => AppSpec.GetDeclForType<DeclType>['declValue']
+) {
+  return [...configs].map(([name, config]) => ({
+    declType: type,
+    declName: name,
+    declValue: configToDeclValue(config),
+  }))
```

Comment:
This might look too complex compared to what we had before (the `for` loops), but comes with many benefits:

#### No repetition
We don't repeat the same map iteration and `Decl` object construction a bunch of times.

#### The entire `decls` array is built at once, without pushing
Because we're functional bros.


#### The compiler ensures we don't forget to include any declarations
This is the biggest benefit IMO, courtesy of the `makeDeclsArray` function.
For example, see what happens if we forget to include Crud:

![image](https://github.com/user-attachments/assets/2a270e92-5ef2-468d-b99d-7057dc2e951c)

So, when we add a new declaration, the compiler tells us what to implement.

#### The compiler ensures we don't mix anything up
`mapToDecls` ensures that the `Decl` returned by the mapping function matches the `DeclType` provided as the second argument, so we can't mix different `Decl` types and values (we had this before, but it's preserved):

![image](https://github.com/user-attachments/assets/ce467ade-1a75-48a7-a7b5-d7ce65b7f3bd)